brave/browser/net/proxy/proxy_config_service_tor.cc

```diff
@@ -4,32 +4,136 @@
+      + ":" + port_));
+}
+
+void ProxyConfigServiceTor::OnTorCrashed(int32_t pid) {
+  LOG(ERROR) << "Tor Process(" << pid << ") Crashed";
+  BrowserThread::PostTask(
+      BrowserThread::PROCESS_LAUNCHER, FROM_HERE,
+      base::Bind(&ProxyConfigServiceTor::LaunchTorProcess,
+                 base::Unretained(this)));
 }
 
 ProxyConfigServiceTor::~ProxyConfigServiceTor() {}
 
+void ProxyConfigServiceTor::LaunchTorProcess() {
+  content::ServiceManagerConnection::GetForProcess()
+    ->GetConnector()
+    ->BindInterface(tor::mojom::kTorServiceName,
+                    &tor_launcher_);
+
+  tor_launcher_.set_connection_error_handler(
+    base::BindOnce(&ProxyConfigServiceTor::OnTorLauncherCrashed,
+                   base::Unretained(this)));
+  std::vector<std::string> args;
+  args.push_back("--ignore-missing-torrc");
+  args.push_back("-f");
+  args.push_back("/nonexistent");
+  args.push_back("--defaults-torrc");
+  args.push_back("/nonexistent");
+  args.push_back("--SocksPort");
+  args.push_back(host_ + ":" + port_);
+  base::FilePath user_data_dir;
+  PathService::Get(chrome::DIR_USER_DATA, &user_data_dir);
+  if (!user_data_dir.empty()) {
+    args.push_back("--DataDirectory");
+    args.push_back(user_data_dir.value());
+  }
+  tor_launcher_->Launch(base::FilePath(tor_path_), args,
+                        base::Bind(&ProxyConfigServiceTor::OnTorLaunched,
+                                   base::Unretained(this)));
+  tor_launcher_->SetCrashHandler(base::Bind(&ProxyConfigServiceTor::OnTorCrashed,
+                        base::Unretained(this)));
+}
+
+void ProxyConfigServiceTor::TorSetProxy(
+    scoped_refptr<brightray::URLRequestContextGetter>
+      url_request_context_getter,
+    const std::string tor_proxy,
+    const std::string tor_path,
+    const bool isolated_storage,
+    const base::FilePath partition_path) {
+  DCHECK_CURRENTLY_ON(BrowserThread::UI);
+  if (!url_request_context_getter || !isolated_storage)
+    return;
+  auto proxy_service = url_request_context_getter->GetURLRequestContext()->
+    proxy_service();
+  // Notice CreateRequestContextForStoragePartition will only be called once
+  // per partition_path so there is no need to cache password per origin
+  std::string origin = partition_path.DirName().BaseName().value();
+  std::unique_ptr<net::ProxyConfigServiceTor>
+    config(new ProxyConfigServiceTor(tor_path, tor_proxy));
+  config->SetUsername(origin);
+  proxy_service->ResetConfigService(std::move(config));
+}
+
+void ProxyConfigServiceTor::OnTorLauncherCrashed() {
+  LOG(ERROR) << "Tor Launcher Crashed";
+}
+
+void ProxyConfigServiceTor::OnTorLaunched(bool result) {
+  if (!result)
+    LOG(ERROR) << "Tor Launching Failed";
+}
+
 ProxyConfigServiceTor::ConfigAvailability
     ProxyConfigServiceTor::GetLatestProxyConfig(ProxyConfig* config) {
-  if (!url_.is_valid())
+  if (scheme_ != kSocksProxy || host_ != kTorHost || port_ != kTorPort)
```

Comment:
Just check the host port string are not empty. Port varies among different channel and in the future it might changes dynamically in order to find usable port.